File: Snooper/src/test/java/com/prateekj/snooper/SnooperShakeListenerTest.java
package com.prateekj.snooper;

import org.junit.Test;
import org.mockito.InOrder;
import org.mockito.Mockito;

import static org.mockito.Mockito.mock;

public class SnooperShakeListenerTest {

  @Test
  public void shouldVerifySnooperShakeBehaviour() throws Exception {
    SnooperShakeAction shakeAction = mock(SnooperShakeAction.class);
    InOrder inOrder = Mockito.inOrder(shakeAction);
    SnooperShakeListener shakeListener = new SnooperShakeListener(shakeAction);

    shakeListener.onShake();
    inOrder.verify(shakeAction).startSnooperFlow();

    shakeListener.onShake();
    inOrder.verify(shakeAction).endSnooperFlow();

    shakeListener.onShake();
    inOrder.verify(shakeAction).startSnooperFlow();

    shakeListener.onShake();
    inOrder.verify(shakeAction).endSnooperFlow();
  }
}

File: Snooper/src/test/java/com/prateekj/snooper/SnooperShakeListenerTest.kt
package com.prateekj.snooper

import io.mockk.mockk
import io.mockk.verifySequence
import org.junit.Test

class SnooperShakeListenerTest {

  @Test
  @Throws(Exception::class)
  fun shouldVerifySnooperShakeBehaviour() {
    val shakeAction = mockk<SnooperShakeAction>(relaxed = true)
    val shakeListener = SnooperShakeListener(shakeAction)

    repeat(4) { shakeListener.onShake() }

    verifySequence {
      shakeAction.startSnooperFlow()
      shakeAction.endSnooperFlow()
      shakeAction.startSnooperFlow()
      shakeAction.endSnooperFlow()
    }
  }
}